addons/project_timesheet_holidays/models/hr_holidays.py

```diff
@@ -56,45 +56,53 @@ class Holidays(models.Model):
-        """ Timesheet will be generated on leave validation only if a timesheet_project_id and a
-            timesheet_task_id are set on the corresponding leave type. The generated timesheet will
-            be attached to this project/task.
+        """ Timesheet will be generated on leave validation only if timesheet_generate is True
+            If company is set, timesheet_project_id and timesheet_task_id from leave type are
+            used as project_id and task_id.
+            Else, internal_project_id and leave_timesheet_task_id are used.
+            The generated timesheet will be attached to this project/task.
         """
-        holidays = self.filtered(
-            lambda l: l.holiday_type == 'employee' and
-            l.holiday_status_id.timesheet_project_id and
-            l.holiday_status_id.timesheet_task_id and
-            l.holiday_status_id.timesheet_project_id.sudo().company_id == (l.holiday_status_id.company_id or self.env.company))
-
-        # Unlink previous timesheets do avoid doublon (shouldn't happen on the interface but meh)
-        old_timesheets = holidays.sudo().timesheet_ids
-        if old_timesheets:
-            old_timesheets.holiday_id = False
-            old_timesheets.unlink()
+        self = self.sudo()
+        old_timesheets = self.env["account.analytic.line"]
+        vals_list = []
+        for leave in self:
+            if leave.holiday_type != 'employee' or not leave.holiday_status_id.timesheet_generate:
+                continue
 
-        # create the timesheet on the vacation project
-        holidays._timesheet_create_lines()
+            if leave.holiday_status_id.company_id:
+                project_id, task_id = leave.holiday_status_id.timesheet_project_id, leave.holiday_status_id.timesheet_task_id
+            else:
+                project_id, task_id = leave.employee_id.company_id.internal_project_id, leave.employee_id.company_id.leave_timesheet_task_id
 
-        return super()._validate_leave_request()
+            if not project_id or not task_id:
+                continue
 
-    def _timesheet_create_lines(self):
-        vals_list = []
-        for leave in self:
+            old_timesheets |= leave.timesheet_ids
             if not leave.employee_id:
                 continue
+
             work_hours_data = leave.employee_id.list_work_time_per_day(
                 leave.date_from,
                 leave.date_to)
+
             for index, (day_date, work_hours_count) in enumerate(work_hours_data):
-                vals_list.append(leave._timesheet_prepare_line_values(index, work_hours_data, day_date, work_hours_count))
-        return self.env['account.analytic.line'].sudo().create(vals_list)
+                vals_list.append(leave._timesheet_prepare_line_values(index, work_hours_data, day_date, work_hours_count, project_id.id, task_id.id))
+
+        # Unlink previous timesheets to avoid doublon (shouldn't happen on the interface but meh)
```

Comment:
```suggestion
        # Unlink previous timesheets to avoid doublon (shouldn't happen on the interface but meh)
        old_timesheets = self.env[...].search([('project_id', '!=', False), ('holiday_id', 'in', leave_ids)])
```